astronomer/providers/google/cloud/operators/dataproc.py

```diff
@@ -271,17 +271,26 @@ def execute(self, context: Context) -> None:
         self.job_id = job_id
         # Save data required for extra links no matter what the job status will be
         DataprocLink.persist(context=context, task_instance=self, url=DATAPROC_JOB_LOG_LINK, resource=job_id)
-        self.defer(
-            timeout=self.execution_timeout,
-            trigger=DataProcSubmitTrigger(
-                gcp_conn_id=self.gcp_conn_id,
-                dataproc_job_id=job_id,
-                project_id=self.project_id,
-                region=self.region,
-                impersonation_chain=self.impersonation_chain,
-            ),
-            method_name="execute_complete",
-        )
+
+        job = self.hook.get_job(project_id=self.project_id, region=self.region, job_id=job_id)
+        state = job.status.state
+        if state not in (JobStatus.State.ERROR, JobStatus.State.DONE, JobStatus.State.CANCELLED):
+            self.defer(
+                timeout=self.execution_timeout,
+                trigger=DataProcSubmitTrigger(
+                    gcp_conn_id=self.gcp_conn_id,
+                    dataproc_job_id=job_id,
+                    project_id=self.project_id,
+                    region=self.region,
+                    impersonation_chain=self.impersonation_chain,
+                ),
+                method_name="execute_complete",
+            )
+        elif state == JobStatus.State.ERROR:
+            raise AirflowException(f"Job failed:\n{job}")
+        elif state == JobStatus.State.CANCELLED:
+            raise AirflowException(f"Job was cancelled:\n{job}")
+        return job_id
```

Comment:
Suggestion to use Early Return approach here. https://medium.com/swlh/return-early-pattern-3d18a41bba8 . Especially the second example on checking exceptions early. This also allows to have less nesting.

It could look something like the below:
```
if state == error:
    raise Exception
elif state == cancelled:
    raise exception
elif state ==done:
    return job_id
self.defer(
    ....,
    ....,
)
```

Reply:
Thanks for your suggestion. I think it makes sense. Just updated it